File: src/mcp/context.ts
/**
 * MCP runtime contract.
 *
 * The built-in MCP servers (agentos, inbox, r2-fs) are product logic exposed
 * as tools. They run in-process for the simulated runner and are mounted into
 * real Claude Agent SDK sessions via createSdkMcpServer (cloud runner). All
 * side effects go through callbacks implemented by the SessionService, which
 * also streams tool calls to the live viewer.
 */
import { z } from "zod";
import type { SessionManifest, InboxMessage, ToolCallLogEntry, Task } from "../domain/types.js";
import type { Services } from "../services/registry.js";
import { networkAllowed } from "../acl/grants.js";

export interface McpRuntime {
  sessionId: string;
  manifest: SessionManifest;
  services: Services;
  /** Persist + stream one tool call. */
  recordToolCall(entry: ToolCallLogEntry): Promise<void>;
  /** Called when an agent asks/sends via inbox: pauses the session. */
  onInboxQuestion(message: InboxMessage): Promise<void>;
  /** Called when an agent sends a fire-and-forget inbox note (no pause). */
  onInboxNote(message: InboxMessage): Promise<void>;
  /** Spawn a collaborator (collaboration list enforced by the caller). */
  spawnCollaborator(agentName: string, brief: string): Promise<Task>;
  /** Current human answer for a resumed inbox.ask (null when none). */
  answer(): { body?: string; selectedChoiceId?: string; label?: string } | null;
  /** Block until the human answers the current inbox.ask (resume). */
  waitForAnswer(): Promise<{ body?: string; selectedChoiceId?: string; label?: string } | null>;
  /** Deliver a human answer to a waiting agent. */
  injectAnswer(answer: { body?: string; selectedChoiceId?: string; label?: string }): void;
  /** Abort an outstanding waitForAnswer (session being destroyed). */
  cancelWait(): void;
}

export interface McpTool<Args extends z.ZodType<any> = z.ZodType<any>> {
  name: string;
  description: string;
  inputSchema: Args;
  handler: (rt: McpRuntime, args: z.infer<Args>) => Promise<unknown>;
}

export interface McpServer {
  name: string;
  tools: McpTool[];
  /**
   * Hosts this server's tools reach over the network (external connections
   * only). The §5.5 network wall is enforced here, at the one dispatch choke
   * point every runner (simulated, local, cloud SDK) goes through: a limited
   * environment denies any host that is not on its allowlist. Built-in
   * product servers (agentos, inbox, r2-fs) are in-process control-plane
   * logic and have no hosts.
   */
  networkHosts: string[];
  /** Call a tool by name; throws Error on unknown tool / zod failure. */
  call(rt: McpRuntime, name: string, args: unknown): Promise<unknown>;
  listTools(): { name: string; description: string; inputSchema: Record<string, unknown> }[];
}

export function createMcpServer<Tools extends McpTool[]>(
  name: string,
  tools: Tools,
  networkHosts: string[] = [],
): McpServer {
  const byName = new Map(tools.map((t) => [t.name, t]));
  return {
    name,
    tools,
    networkHosts,
    async call(rt, toolName, args) {
      const tool = byName.get(toolName);
      if (!tool) throw new Error(`unknown tool ${name}.${toolName}`);
      // §5.5 / §21 Phase 2 network wall: the environment policy is enforced
      // server-side before any external connection's tool may run.
      for (const host of networkHosts) {
        if (!networkAllowed(rt.manifest.environment, host)) {
          throw new Error(
            `network wall denied: ${name} (${host}) is not in the environment allowlist`,
          );
        }
      }
      const parsed = tool.inputSchema.safeParse(args ?? {});
      if (!parsed.success) {
        throw new Error(`invalid args for ${name}.${toolName}: ${parsed.error.message}`);
      }
      return tool.handler(rt, parsed.data);
    },
    listTools() {
      return tools.map((t) => ({
        name: t.name,
        description: t.description,
        inputSchema: jsonSchemaOf(t.inputSchema),
      }));
    },
  };
}

/** Convert a zod schema to a JSON-schema-ish object for SDK MCP registration. */
export function jsonSchemaOf(schema: z.ZodType): Record<string, unknown> {
  try {
    return z.toJSONSchema(schema) as Record<string, unknown>;
  } catch {
    return { type: "object" };
  }
}

export function callToolWithLog(
  server: McpServer,
  rt: McpRuntime,
  name: string,
  args: unknown,
): Promise<unknown> {
  return server.call(rt, name, args);
}

File: agentos/mcp/context.py
"""
MCP runtime context. Port of src/mcp/context.ts.
"""
from __future__ import annotations

import asyncio
from typing import Any, Callable, Optional


class McpRuntime:
    def __init__(self, *, session_id: str, manifest: Any, services: Any,
                 on_tool_call: Callable, on_status: Callable) -> None:
        self.session_id = session_id
        self.manifest = manifest
        self.services = services
        self._on_tool_call = on_tool_call
        self._on_status = on_status
        self._answer: Optional[dict] = None
        self._answer_event: asyncio.Event = asyncio.Event()
        self._cancelled = False

    async def record_tool_call(self, entry: dict) -> None:
        await self._on_tool_call(self.session_id, entry)

    async def on_inbox_question(self, message: dict) -> None:
        await self._on_status(self.session_id, "waiting-inbox")

    async def on_inbox_note(self, message: dict) -> None:
        pass  # fire-and-forget; session keeps running

    async def spawn_collaborator(self, agent_name: str, brief: str) -> dict:
        list_ = self.manifest.collaborationList
        if agent_name not in list_:
            raise ValueError(f'collaborator "{agent_name}" not on list [{", ".join(list_)}]')
        from agentos.db.models import Agent as AgentRow
        from agentos.db.client import get_session as _get_session
        from sqlalchemy import select, and_
        async with _get_session() as db:
            result = await db.execute(
                select(AgentRow).where(
                    and_(AgentRow.project_id == self.manifest.projectId,
                         AgentRow.name == agent_name)
                )
            )
            agent = result.scalar_one_or_none()
        if not agent:
            raise ValueError(f'unknown agent "{agent_name}"')
        return await self.services.tasks.spawn_collaborator(
            project_id=self.manifest.projectId,
            agent_id=agent.id,
            agent_name=agent_name,
            brief=brief,
            parent_task_id=self.manifest.task.id if self.manifest.task else None,
            parent_session_id=self.session_id,
        )

    def answer(self) -> Optional[dict]:
        return self._answer

    async def wait_for_answer(self) -> Optional[dict]:
        self._answer_event.clear()
        self._answer = None
        await self._answer_event.wait()
        if self._cancelled:
            return None
        return self._answer

    def inject_answer(self, answer: dict) -> None:
        self._answer = answer
        self._answer_event.set()

    def cancel_wait(self) -> None:
        self._cancelled = True
        self._answer_event.set()


class McpServer:
    def __init__(self, name: str, tools: list["McpTool"], network_hosts: list[str] = None) -> None:
        self.name = name
        self.tools = tools
        self.network_hosts = network_hosts or []
        self._by_name = {t.name: t for t in tools}

    async def call(self, rt: McpRuntime, tool_name: str, args: Any) -> Any:
        tool = self._by_name.get(tool_name)
        if not tool:
            raise ValueError(f"unknown tool {self.name}.{tool_name}")
        # Network wall
        from agentos.acl.grants import network_allowed
        for host in self.network_hosts:
            if not network_allowed(rt.manifest.environment, host):
                raise PermissionError(
                    f"network wall denied: {self.name} ({host}) not in environment allowlist"
                )
        return await tool.handler(rt, args or {})

    def list_tools(self) -> list[dict]:
        return [{"name": t.name, "description": t.description} for t in self.tools]


class McpTool:
    def __init__(self, name: str, description: str, handler: Callable) -> None:
        self.name = name
        self.description = description
        self.handler = handler


def create_mcp_server(name: str, tools: list[McpTool], network_hosts: list[str] = None) -> McpServer:
    return McpServer(name, tools, network_hosts or [])
